pkg/provider/aws/client.go

```diff
@@ -166,344 +159,326 @@ func addProxyConfigToSessionOptConfig(config *aws.Config) {
 		if err != nil {
-			return nil, fmt.Errorf("could not load config file: %v", err)
+			return nil, fmt.Errorf("could not load config file: %w", err)
+		}
+		cfg, err = config.LoadDefaultConfig(context.TODO(), config.WithRegion(region), config.WithSharedConfigProfile(profile), config.WithSharedConfigFiles([]string{absCfgPath}))
+	} else {
+		cfg, err = config.LoadDefaultConfig(context.TODO(), config.WithRegion(region), config.WithSharedConfigProfile(profile))
+		if err != nil {
+			return nil, fmt.Errorf("error loading aws config: %w", err)
 		}
-		opt.SharedConfigFiles = []string{absCfgPath}
 	}
 
-	sess := session.Must(session.NewSessionWithOptions(opt))
-	if _, err := sess.Config.Credentials.Get(); err != nil {
-		if aerr, ok := err.(awserr.Error); ok {
-			switch aerr.Code() {
-			case "NoCredentialProviders":
-				return nil, fmt.Errorf("could not create AWS session: %v", err)
-			default:
-				return nil, fmt.Errorf("could not create AWS session: %v", err)
-			}
+	addProxyConfigToSessionOptConfig(&cfg)
+
+	if _, err := cfg.Credentials.Retrieve(context.TODO()); err != nil {
+		if err != nil {
+			return nil, fmt.Errorf("failed to retrieve AWS credentials: %w", err)
 		}
 	}
 
-	return sess, nil
+	return &cfg, nil
 }
 
 // NewAwsClient creates an AWS client with credentials in the environment
 func NewAwsClient(profile, region, configFile string) (Client, error) {
-	sess, err := NewAwsSession(profile, region, configFile)
+	cfg, err := NewAwsConfig(profile, region, configFile)
 	if err != nil {
 		return nil, err
 	}
 
 	awsClient := &AwsClient{
-		iamClient:           iam.New(sess),
-		ec2Client:           ec2.New(sess),
-		stsClient:           sts.New(sess),
-		s3Client:            s3.New(sess),
-		servicequotasClient: servicequotas.New(sess),
-		orgClient:           organizations.New(sess),
-		ceClient:            costexplorer.New(sess),
-		resClient:           resourcegroupstaggingapi.New(sess),
-		cloudTrailClient:    cloudtrail.New(sess),
+		iamClient:           *iam.NewFromConfig(*cfg),
+		ec2Client:           *ec2.NewFromConfig(*cfg),
+		stsClient:           *sts.NewFromConfig(*cfg),
+		s3Client:            *s3.NewFromConfig(*cfg),
+		servicequotasClient: *servicequotas.NewFromConfig(*cfg),
+		orgClient:           *organizations.NewFromConfig(*cfg),
+		ceClient:            *costexplorer.NewFromConfig(*cfg),
+		resClient:           *resourcegroupstaggingapi.NewFromConfig(*cfg),
+		cloudTrailClient:    *cloudtrail.NewFromConfig(*cfg),
 	}
 
 	// Validate the creds
 	if _, err := awsClient.GetCallerIdentity(nil); err != nil {
-		if aerr, ok := err.(awserr.Error); ok {
-			switch aerr.Code() {
-			case "InvalidClientTokenId":
-				return nil, fmt.Errorf("`aws sts get-caller-identity --profile %s --region %s` failed - "+
-					"ensure the credentials in the profile are valid and the region is in the AWS partition you expect", profile, region)
-			default:
-				return nil, err
-			}
+		var iite *stsTypes.InvalidIdentityTokenException
```

Comment:
InvalidIdentityTokenException != InvalidClientTokenId

Reply:
The new sts SDK doesn't seem to have an error type of "InvalidClientTokenId", and this was the only error I could find relating to tokens at all so I assumed it was what they renamed it to. The doc for InvalidIdentityTokenException reads:
```
// The web identity token that was passed could not be validated by Amazon Web
// Services. Get a new identity token from the identity provider and then retry the
// request.
```

Any thoughts?

Reply:
Yeah, AWS doesn't have available types for all of their errors, in this case the two errors are not the same. `InvalidClientTokenId` is a common error to STS (https://docs.aws.amazon.com/STS/latest/APIReference/API_GetCallerIdentity.html#API_GetCallerIdentity_Errors) and can happen during `aws sts get-caller-identity` if the AWS Access Key ID is invalid, but `InvalidIdentityToken` specifically only happens during `aws sts assume-role-with-web-identity` https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRoleWithWebIdentity.html#API_AssumeRoleWithWebIdentity_Errors

Reply:
Thanks for the insight. In this case, I'm going to drop the special handling of this particular alert and return the error through from the AWS SDK.